Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1541,17 +1541,23 @@ extension BrowserViewController: URLBarDelegate {
     }
 
     func urlBar(urlBar: URLBarView, didSubmitText text: String) {
-        // If we can't make a valid URL, do a search query.
-        // If we still don't have a valid URL, something is broken. Give up.
         let engine = profile.searchEngines.defaultEngine
-        guard let url = URIFixup.getURL(text) ??
-                        engine.searchURLForQuery(text) else {
+        let url: NSURL
+
+        if let fixupURL = URIFixup.getURL(text) {
+            // The user entered a URL, so use it.
+            url = fixupURL
+        } else if let searchURL = engine.searchURLForQuery(text) {
+            // We couldn't build a URL, so do a search query.
+            url = searchURL
+            Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        } else {
+            // We still don't have a valid URL, so something is broken. Give up.
             log.error("Error handling URL entry: \"\(text)\".")
```

Comment:
My only concern with baking in the assert with the log.error is that I don't think it's always the case that we want to assert if we use log.error. It's not clear from the calling standpoint that calling it will throw an assert. I think the asserts should be explicit in their use.